THEORY QUESTIONS:

Ans1.

In Python, functions and methods are both callable pieces of code, but they have some key differences based on how they are defined and used:

1. Definition

.Function: A function is a standalone block of code defined using the def keyword. It can be called independently of any object.

In [1136]:
def greet(name):
    return f"Hello, {name}!"

.Method: A method is a function that is associated with a class or an object. It is defined within a class and is called on an instance (or the class itself for class methods).

In [1137]:
class Greeter:
    def greet(self, name):
        return f"Hello, {name}!"

2.Calling Context

.Function: invoked directly,without requiring an intance or class.

In [1138]:
print(greet("Alice"))  # Output: Hello, Alice!

Hello, Alice!


.Method: Called on an instance (or class, in the case of class methods). The first parameter of an instance method is typically self, representing the instance, while for class methods it is cls.

In [1139]:
greeter = Greeter()
print(greeter.greet("Alice"))  # Output: Hello, Alice!

Hello, Alice!


3. Binding

.Function: Not bound to any object or class. It does not have implicit access to instance or class-specific data.

.Method: Bound to an instance (or class). Instance methods have access to the object's attributes via self, and class methods can access the class-level data via cls.

4. Usage Context

.Function: Used for general-purpose operations that do not specifically depend on object state.

.Method: Used when operations are tied to an object's state or behavior.


Special Case: Static Methods

Python also has static methods, defined with the @staticmethod decorator, which behave like regular functions but are included within a class for logical grouping. They do not take self or cls as the first argument:

In [1140]:
class Utils:
    @staticmethod
    def add(a, b):
        return a + b

print(Utils.add(2, 3))  # Output: 5

5


Ans 2.

In Python, arguments and parameters are closely related concepts used in the context of functions. While they are often used interchangeably, they refer to distinct aspects of how data is passed to functions.

1. Parameters

.Definition: Parameters are the variables listed in a function's definition. They act as placeholders for the values that will be passed to the function when it is called.

.Purpose: Parameters define the inputs that the function expects to receive.

In [1141]:
def greet(name):  # 'name' is a parameter
    return f"Hello, {name}!"

2. Arguments

.Definition: Arguments are the actual values or data you pass to a function when you call it.

.Purpose: Arguments provide the specific data that the function will use.

Example:

In [1142]:
print(greet("Alice"))  # "Alice" is an argument

Hello, Alice!


Types of Function Arguments and Parameters

Python supports several types of parameters and arguments, allowing flexible function definitions and calls:

1. Positional Arguments

.Passed in order, matching the position of parameters.

In [1143]:
def add(a, b):
    return a + b

print(add(2, 3))  # 2 and 3 are positional arguments

5


2. Keyword Arguments

.Explicitly specify which parameter the argument corresponds to, using the parameter name.

In [1144]:
def greet(name, message):
    return f"{message}, {name}!"

print(greet(name="Alice", message="Hi"))  # Keyword arguments

Hi, Alice!


3. Default Parameters

.Assign default values to parameters. These are used if no argument is passed for them.

In [1145]:
def greet(name, message="Hello"):
    return f"{message}, {name}!"

print(greet("Alice"))  # Uses default message: "Hello"

Hello, Alice!


4. Variable-Length Arguments

.*args: Collects additional positional arguments as a tuple.

In [1146]:
def add(*numbers):
    return sum(numbers)

print(add(1, 2, 3, 4))  # Output: 10

10


.**kwargs: Collects additional keyword arguments as a dictionary.

In [1147]:
def info(**details):
    return details

print(info(name="Alice", age=25))  # Output: {'name': 'Alice', 'age': 25}

{'name': 'Alice', 'age': 25}


Ans 3.

In Python, functions can be defined and called in various ways to support diverse programming requirements. Here's an overview of the different ways to define and call functions:

1. Basic Function Definition and Call

Definition:

A simple function is defined using the def keyword, followed by the function name and parentheses containing parameters (if any).

In [1148]:
def greet(name):
    return f"Hello, {name}!"

Call:

In [1149]:
print(greet("Alice")) #output: Hello, Alice!

Hello, Alice!


2. Function with Default Parameters

Definition:

Parameters can have default values, which are used if no argument is provided during the call.

In [1150]:
def greet(name, message="Hello"):
    return f"{message}, {name}!"

Call:

In [1151]:
print(greet("Alice"))            # Uses default: "Hello, Alice!"
print(greet("Bob", "Hi"))        # Output: Hi, Bob!

Hello, Alice!
Hi, Bob!


3. Function with Variable-Length Arguments

Definition:

.*args: Captures extra positional arguments as a tuple.

.**kwargs: Captures extra keyword arguments as a dictionary.

In [1152]:
def describe(*args, **kwargs):
    return f"Positional: {args}, Keyword: {kwargs}"

Call:

In [1153]:
print(describe(1, 2, 3, name="Alice", age=25))  
# Output: Positional: (1, 2, 3), Keyword: {'name': 'Alice', 'age': 25}

Positional: (1, 2, 3), Keyword: {'name': 'Alice', 'age': 25}


4. Lambda (Anonymous) Functions

Definition:

Lambda functions are single-expression functions defined using the lambda keyword. They are commonly used for short, throwaway functions.

In [1154]:
add = lambda x, y: x + y

Call:

In [1155]:
print(add(2,3)) #output: 5

5


5. Nested Functions

Definition:

Functions can be defined within other functions. This is useful for encapsulating functionality.

In [1156]:
def outer_function(msg):
    def inner_function():
        return f"Inner says: {msg}"
    return inner_function()

Call:

In [1157]:
print(outer_function("Hi there!"))  # Output: Inner says: Hi there!

Inner says: Hi there!


6. Recursive Functions

Definition:

A function that calls itself to solve smaller sub-problems, typically used for tasks like factorial or Fibonacci series.

In [1158]:
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n - 1)

Call:

In [1159]:
print(factorial(5))  #Output :120

120


7. Function with Type Hints

Definition:

Type hints specify the expected input and output types for better readability and debugging.

In [1160]:
def add(x: int, y: int) -> int:
    return x + y

Call:

In [1161]:
print(add(2,3))  #Output: 5

5


Ans 8.

A lambda function in Python is a small, anonymous function that can have any number of arguments but only one expression. It is defined using the lambda keyword. The expression is evaluated and returned when the function is called.

lambda arguments: expression

.lambda – Keyword to define the function.
.arguments – Input parameters.
.expression – Single line of code that returns a value.

In [1162]:
add = lambda x, y: x + y
print(add(5, 3))  # Output: 8

8


In [1163]:
def add(x, y):
    return x + y

When to Use Lambda Functions:

1. Short, Simple Operations

Lambda functions are used for small, throwaway functions where defining a full function is unnecessary.

2. Higher-Order Functions

Functions like map(), filter(), and reduce() often use lambdas.

In [1164]:
numbers = [1, 2, 3, 4]
squared = map(lambda x: x**2, numbers)
print(list(squared))  

[1, 4, 9, 16]


3. Sorting with Custom Keys

In [1165]:
names = ['John', 'Alex', 'Emily']
names.sort(key=lambda x: len(x))
print(names)  

['John', 'Alex', 'Emily']


4. Conditionally Selecting Elements

In [1166]:
even_numbers = filter(lambda x: x % 2 == 0, range(10))
print(list(even_numbers))  

[0, 2, 4, 6, 8]


5. Inline Callbacks
Useful in GUI programming or asynchronous operations where small callbacks are required.

Ans 9.

The map() function in Python applies a given function to all items in an iterable (like a list or tuple) and returns an iterator (or list in Python 2) with the results. It allows for efficient, element-wise transformations without the need for explicit loops.

map(function, iterable, ...)

.function – A function that is applied to each element.
.iterable – One or more iterables (like lists, tuples).
If multiple iterables are passed, the function must accept that many arguments.

Example 1: Basic Usage (Single Iterable)

In [1167]:
numbers = [1, 2, 3, 4]
squared = map(lambda x: x**2, numbers)
print(list(squared))  

[1, 4, 9, 16]


Explanation:

.lambda x: x**2 squares each element in the list.
.map applies this to every item in numbers.

Example 2 : Multiple Iterables

In [1168]:
list1 = [1, 2, 3]
list2 = [4, 5, 6]
summed = map(lambda x, y: x + y, list1, list2)
print(list(summed))  

[5, 7, 9]


Explanation:

.Each element of list1 is summed with the corresponding element of list2.

Example 3: Using a Named Function

In [1169]:
def to_uppercase(s):
    return s.upper()

words = ['hello', 'world']
uppercase_words = map(to_uppercase, words)
print(list(uppercase_words))  

['HELLO', 'WORLD']


Explanation:

.map works with regular functions, not just lambdas.

Ans 10.

map(), reduce(), and filter() are functional programming tools in Python that operate on iterables. They allow you to apply functions efficiently without explicit loops. Here's a breakdown of their differences and use cases:

1. map() – Apply a Function to Each Item
.Purpose: Transforms each element in an iterable by applying a function.
.Returns: An iterator with the transformed elements.
.Key Use Case: Element-wise transformations.
.Syntax:

map(function, iterable, ...)

Example:

In [1170]:
numbers = [1, 2, 3, 4]
squared = map(lambda x: x**2, numbers)
print(list(squared))  

[1, 4, 9, 16]


2. filter() – Select Items Based on a Condition
.Purpose: Filters elements from an iterable based on a condition (Boolean function).
.Returns: An iterator with elements that satisfy the condition (where the function returns True).
.Key Use Case: Extracting items that meet certain criteria.
.Syntax:

filter(function, iterable)

Example:

In [1171]:
numbers = [1, 2, 3, 4, 5]
even_numbers = filter(lambda x: x % 2 == 0, numbers)
print(list(even_numbers))  

[2, 4]


3. reduce() – Cumulatively Apply a Function to Reduce to a Single Value

.Purpose: Repeatedly applies a function to pairs of elements, reducing the iterable to a single cumulative value.
.Returns: A single value.
.Key Use Case: Aggregation (e.g., sum, product).
.Syntax:

from functools import reduce
reduce(function, iterable, initial_value)

In [1172]:
from functools import reduce
numbers = [1, 2, 3, 4]
product = reduce(lambda x, y: x * y, numbers)
print(product)  

24


When to Use Each:

.map() – When you need to transform each item in a list.
.filter() – When you need to select certain items based on a condition.
.reduce() – When you need to combine elements into a single cumulative value.

Example(All Together):

In [1173]:
numbers = [1, 2, 3, 4, 5]

# 1. Double each number (map)
doubled = map(lambda x: x * 2, numbers)

# 2. Filter even numbers (filter)
even_numbers = filter(lambda x: x % 2 == 0, doubled)

# 3. Sum the remaining numbers (reduce)
total = reduce(lambda x, y: x + y, even_numbers)

print(total)  

30


Practical Questions:

Ans 1.

In [1174]:
def sum_even_numbers(numbers):
    return sum(filter(lambda x: x % 2 == 0, numbers))

# Example usage
numbers = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(numbers)
print(result)  

12


How it Works:

.filter() – Filters out even numbers by applying the condition x % 2 == 0.
.sum() – Adds up the filtered even numbers.

Alternative Approach (Using List Comprehension):

In [1175]:
def sum_even_numbers(numbers):
    return sum(x for x in numbers if x % 2 == 0)

# Example usage
print(sum_even_numbers([1, 2, 3, 4, 5, 6]))  

12


Ans 2.

In [1176]:
def reverse_string(s):
    return s[::-1]

# Example usage
text = "hello"
result = reverse_string(text)
print(result)  

olleh


How it Works:

.s[::-1] – Uses Python slicing to reverse the string:
.[:] – Selects the entire string.
.[::-1] – Steps backward by -1, effectively reversing the string.

Alternative Approach (Using a Loop):

In [1177]:
def reverse_string(s):
    reversed_str = ''
    for char in s:
        reversed_str = char + reversed_str
    return reversed_str

# Example usage
print(reverse_string("world"))  

dlrow


Alternative Approach (Using reversed() and join()):

In [1178]:
def reverse_string(s):
    return ''.join(reversed(s))

# Example usage
print(reverse_string("Python"))  

nohtyP


Ans 3.

In [1179]:
def square_numbers(numbers):
    return [x**2 for x in numbers]

# Example usage
nums = [1, 2, 3, 4, 5]
result = square_numbers(nums)
print(result)  

[1, 4, 9, 16, 25]


How it Works:

.List Comprehension – Squares each element (x**2) and creates a new list.
.Efficient and Readable – This approach is concise and Pythonic.

Alternative Approach (Using map()):

In [1180]:
def square_numbers(numbers):
    return list(map(lambda x: x**2, numbers))

# Example usage
print(square_numbers([1, 2, 3, 4, 5]))  

[1, 4, 9, 16, 25]


Ans 4.

In [1181]:
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Example usage: Check primes from 1 to 200
prime_numbers = [n for n in range(1, 201) if is_prime(n)]
print(prime_numbers)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199]


How it Works:

.n <= 1 – Numbers less than or equal to 1 are not prime.
.Loop up to √n – A number is prime if no divisors exist up to its square root. 
.This reduces the number of iterations, making it efficient.
.range(2, √n + 1) – Checks divisibility from 2 to the square root of n.
.List Comprehension – Generates prime numbers between 1 and 200.

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47,
 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107,
 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167,
 173, 179, 181, 191, 193, 197, 199]

Ans 5.

In [1182]:
class FibonacciIterator:
    def __init__(self, n_terms):
        self.n_terms = n_terms
        self.current = 0
        self.next = 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.n_terms:
            raise StopIteration
        if self.count == 0:
            self.count += 1
            return 0
        elif self.count == 1:
            self.count += 1
            return 1
        else:
            fib = self.current + self.next
            self.current, self.next = self.next, fib
            self.count += 1
            return self.current

# Example usage
fib = FibonacciIterator(10)
print(list(fib))  

[0, 1, 1, 1, 2, 3, 5, 8, 13, 21]


How it Works:

.__init__ – Initializes the iterator with the number of terms (n_terms), starting values for Fibonacci (0 and 1), and a counter.
.__iter__ – Returns the iterator object (self).
.__next__ – Generates the next Fibonacci number:
.Stops after n_terms by raising StopIteration.
.Uses Fibonacci logic (current + next) to compute the next term.

Example Output (10 terms):

In [1183]:
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

Ans 6.

In [1184]:
def powers_of_two(max_exponent):
    for i in range(max_exponent + 1):
        yield 2 ** i

# Example usage
for power in powers_of_two(5):
    print(power)

1
2
4
8
16
32


How it Works:

.yield – Generates powers of 2 one at a time, pausing after each yield and resuming on the next call.
.range(max_exponent + 1) – Iterates from 0 to the specified exponent, inclusive.
.2 ** i – Computes 2 raised to the power of i.

Example Output (up to 2^5):

In [1185]:
1  # 2^0
2  # 2^1
4  # 2^2
8  # 2^3
16 # 2^4
32 # 2^5

32

Ans 7.

def read_file_line_by_line(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line.strip()  # strip removes the newline character

# Example usage
file_path = 'sample.txt'  # Replace with your file path
for line in read_file_line_by_line(file_path):
    print(line)

    How it Works:

.with open(file_path, 'r') – Opens the file in read mode and ensures it is properly closed after reading.
.for line in file – Iterates over each line in the file.
.yield line.strip() – Yields each line, stripping the newline character at the end of each line using .strip().

.Benefits of Using a Generator:

.Memory Efficient – Since the generator yields one line at a time, it doesn't load the entire file into memory, making it efficient for large files.
.Lazy Evaluation – The lines are read and processed on-demand, improving performance in situations where you don't need all the data at once.

Example File Content (sample.txt):

Hello, world!
This is a sample file.
It has multiple lines.

Example Output:

Hello, world!
This is a sample file.
It has multiple lines.

Ans 8.

You can use a lambda function with the sorted() function in Python to sort a list of tuples based on the second element of each tuple. Here's an example:

In [1186]:
# List of tuples
tuples = [(1, 'apple'), (3, 'banana'), (2, 'cherry')]

# Sort the list of tuples based on the second element (index 1) of each tuple
sorted_tuples = sorted(tuples, key=lambda x: x[1])

# Output the sorted list
print(sorted_tuples)

[(1, 'apple'), (3, 'banana'), (2, 'cherry')]


How it Works:

.lambda x: x[1] – Defines an anonymous function that returns the second element (x[1]) of each tuple.
.sorted(tuples, key=...) – The sorted() function sorts the list of tuples based on the value returned by the lambda function (i.e., the second element of each tuple).
Example Output:

In [1187]:
[(3, 'banana'), (1, 'apple'), (2, 'cherry')]

[(3, 'banana'), (1, 'apple'), (2, 'cherry')]